File: Lab9-NSNotification/Lab9-NSNotification/Lab9-NSNotification/ThirdVC/ThirdVC.m
//
//  ThirdVCVC.m
//  Lab9-NSNotification
//
//  Created by 張力元 on 2019/3/19.
//  Copyright © 2019 張力元. All rights reserved.
//

#import "ThirdVC.h"

@interface ThirdVC ()
@property (strong, nonatomic) IBOutlet UILabel *sender;
@property (strong, nonatomic) IBOutlet UILabel *msg;
@property (strong, nonatomic) IBOutlet UILabel *timeLabel;
@end

@implementation ThirdVC

- (void)viewDidLoad {
    [super viewDidLoad];
    //註冊區域通知
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(getRegionMsg:) name:@"regionNotification" object:nil];
    //註冊全域通知
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(getPublicMsg:) name:@"publicNotification" object:nil];
}

- (void)getRegionMsg:(NSNotification *)notification{
    NSDictionary *msgData = notification.object;
    [self showMsg:msgData];
}

- (void)showMsg:(NSDictionary *)msgData{
    _sender.text = [msgData valueForKey:@"sender"];
    _msg.text = [msgData valueForKey:@"msg"];
}

- (void)getPublicMsg:(NSNotification *)notification{
    NSDictionary *timeData = notification.object;
    [self showTime:timeData];
}

- (void)showTime:(NSDictionary *)timeData{
    int hour = [[timeData valueForKey:@"hour"] intValue];
    int min = [[timeData valueForKey:@"min"] intValue];
    int sec = [[timeData valueForKey:@"sec"] intValue];
    _timeLabel.text = [NSString stringWithFormat:@"%.2d:%.2d:%.2d",hour,min,sec];
}


@end

File: Swift/lab9/lab9/lab9/thirdVC.swift
//
//  thirdVC.swift
//  lab9
//
//  Created by taizhou on 2019/7/30.
//  Copyright © 2019 taizhou. All rights reserved.
//

import UIKit

class thirdVC: UIViewController {
    @IBOutlet var sender: UILabel!
    @IBOutlet var msg: UILabel!
    @IBOutlet var timeLabel: UILabel!
    var publicNotification : NSObjectProtocol?
    var regionNotification : NSObjectProtocol?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        initNotification()
    }
    
    func initNotification(){
        let publicName = NSNotification.Name("publicNotification")
        publicNotification = NotificationCenter.default.addObserver(forName: publicName, object: nil, queue: OperationQueue.main){ (notification) in
            let timeData = notification.object as! Dictionary<String,Int>
            self.showTime(timeData)
        }
        let regionName = NSNotification.Name("regionNotification")
        regionNotification = NotificationCenter.default.addObserver(forName: regionName, object: nil, queue: OperationQueue.main){ (notification) in
            let msgData = notification.object as! Dictionary<String,String>
            self.showMsg(msgData)
        }
    }
     
    func showTime(_ timeData:Dictionary<String,Int>){
        let hour = timeData["hour"]
        let min = timeData["min"]
        let sec = timeData["sec"]
        timeLabel.text = String(format:"%02d:%02d:%02d",hour ?? 0,min ?? 0,sec ?? 0)
    }
    
    func showMsg(_ msgData:Dictionary<String,String>){
        sender.text = msgData["sender"]
        msg.text = msgData["msg"]
    }
}

